G1ANT.Addon.Azure/AzureKeyVaultManager.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System;
+using System.Threading.Tasks;
+using G1ANT.Addon.Azure.Extensions;
+using Microsoft.Azure.KeyVault;
+using Microsoft.Azure.KeyVault.Models;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace G1ANT.Addon.Azure
+{
+    public class AzureKeyVaultManager
+    {
+        private string clientId;
+        private string clientSecret;
+        private int azureTimeout;
+        private Uri keyVaultUri;
+
+        public AzureKeyVaultManager(string clientId, string clientSecret, Uri keyVaultUri, int azureTimeout)
+        {
+            this.clientId = clientId;
+            this.clientSecret = clientSecret;
+            this.azureTimeout = azureTimeout;
+            this.keyVaultUri = keyVaultUri;
+
+            if(!DoesUriEndWithSlash())
+            {
+                throw new Exception("Key Vault Url should end with '/'");
+            }
+
+            ValidateKeyVaultClient().Wait();
+        }
+
+        private bool DoesUriEndWithSlash()
+        {
+            return keyVaultUri.ToString().EndsWith("/");
+        }
+
+        private async Task<string> GetToken(string authority, string resource, string scope)
+        {
+            var authContext = new AuthenticationContext(authority);
+            var clientCred = new ClientCredential(clientId, clientSecret);
+            var result = await authContext.AcquireTokenAsync(resource, clientCred);
+
+            if (result.AccessToken == null)
+            {
+                throw new Exception("Failed to obtain the JWT token");
+            }
+
+            return result.AccessToken;
+        }
+
+        public async Task<string> GetSecret(string secretName)
+        {
+            using (var keyVaultClient = new KeyVaultClient(GetToken))
```

Comment:
IMHO `KeyVaultClient` shouldbe encapsulated by a separate class in order to allow unit testing (just like all IO operations).